src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,545 @@
+			=> session.TransactionContext?.IsInActiveTransaction ?? false;
+
+		/// <inheritdoc />
+		public override void ExecuteWorkInIsolation(ISessionImplementor session, IIsolatedWork work, bool transacted)
+		{
+			using (var tx = new TransactionScope(TransactionScopeOption.Suppress))
+			{
+				base.ExecuteWorkInIsolation(session, work, transacted);
+				tx.Complete();
+			}
+		}
+
+		/// <summary>
+		/// Transaction context for enlisting a session with a system <see cref="System.Transactions.Transaction"/>.
+		/// It is meant for being the concrete class enlisted in the transaction.
+		/// </summary>
+		public class SystemTransactionContext : ITransactionContext, IEnlistmentNotification
+		{
+			/// <summary>
+			/// The transaction in which this context is enlisted.
+			/// </summary>
+			protected internal virtual System.Transactions.Transaction AmbientTransaction { get; }
+			/// <inheritdoc />
+			public bool ShouldCloseSessionOnSystemTransactionCompleted { get; set; }
+			/// <inheritdoc />
+			public bool IsInActiveTransaction { get; protected set; } = true;
+			/// <inheritdoc />
+			public virtual bool CanFlushOnSystemTransactionCompleted => _useConnectionOnSystemTransactionEvents;
+
+			private readonly ISessionImplementor _sessionImplementor;
+			private readonly bool _useConnectionOnSystemTransactionEvents;
+			private readonly System.Transactions.Transaction _originalTransaction;
+			private readonly ManualResetEventSlim _lock = new ManualResetEventSlim(true);
+			private volatile bool _needCompletionLocking = true;
+			// Required for not locking the second/completion phase itself when locking session usages from concurrent threads.
+			private readonly AsyncLocal<bool> _bypassWait = new AsyncLocal<bool>();
+
+			/// <summary>
+			/// Default constructor.
+			/// </summary>
+			/// <param name="sessionImplementor">The session to enlist with the transaction.</param>
+			/// <param name="transaction">The transaction into which the context will be enlisted.</param>
+			/// <param name="useConnectionOnSystemTransactionEvents">See <see cref="Cfg.Environment.UseConnectionOnSystemTransactionEvents"/>.</param>
+			public SystemTransactionContext(
+				ISessionImplementor sessionImplementor,
+				System.Transactions.Transaction transaction,
+				bool useConnectionOnSystemTransactionEvents)
+			{
+				_sessionImplementor = sessionImplementor;
+				_originalTransaction = transaction;
+				var clone = transaction.Clone();
+				AmbientTransaction = clone;
+				clone.TransactionCompleted += TransactionCompleted;
+				_useConnectionOnSystemTransactionEvents = useConnectionOnSystemTransactionEvents;
+			}
+
+			/// <inheritdoc />
+			public virtual void Wait()
+			{
+				if (_isDisposed)
+					return;
+
+				if (_needCompletionLocking && GetAmbientTransactionStatus() != TransactionStatus.Active)
+				{
+					// Rollback case may end the transaction without a prepare phase, apply the lock.
+					Lock();
+				}
+
+				if (_bypassWait.Value)
```

Comment:
If that one fails, that would cause application freeze. Current transaction completion does not trigger it by chance: `AfterTransaction` events currently do not causes any `Wait` call, and even the session disposal from scope has its special dedicated method by-passing the `Wait` in dispose method.

But relying on never having any `Wait` someday called by those operations would be quite a brittle pattern, leading to unexpected freezes on just a small seemingly unrelated change having caused a `Wait` being call from transaction completion.

I am going to add a test for calling the `Wait` from an interceptor just to showcase this.